From a6e44064f5850f703f97ac2444e624b38cd81001 Mon Sep 17 00:00:00 2001
From: Joe Groocock <me@frebib.net>
Date: Sun, 12 Jul 2020 10:30:49 +0100
Subject: HACK: always sync pillar extmods before loading

This is required when the saltenv points to a gitfs backend and the
extmods haven't been sync'ed before running. It would be good if Salt
could do this automatically when running against a new saltenv, and if
it kept the extmods cache separate for each saltenv to prevent
potentially racy behaviour.

Signed-off-by: Joe Groocock <me@frebib.net>
---
 salt/pillar/__init__.py | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/salt/pillar/__init__.py b/salt/pillar/__init__.py
index 57f4b4d367..1d8ba6833a 100644
--- a/salt/pillar/__init__.py
+++ b/salt/pillar/__init__.py
@@ -22,6 +22,7 @@ import salt.utils.cache
 import salt.utils.crypt
 import salt.utils.data
 import salt.utils.dictupdate
+import salt.utils.extmods
 import salt.utils.url
 from salt.exceptions import SaltClientError
 from salt.ext import six
@@ -531,6 +532,9 @@ class Pillar:
         if opts.get("pillar_source_merging_strategy"):
             self.merge_strategy = opts["pillar_source_merging_strategy"]
 
+        # Sync ext_pillar modules before loading them
+        salt.utils.extmods.sync(self.opts, "pillar", saltenv=saltenv)
+
         self.ext_pillars = salt.loader.pillars(ext_pillar_opts, self.functions)
         self.ignored_pillars = {}
         self.pillar_override = pillar_override or {}
-- 
2.30.0

